// @dlr-eoc/layer-control/esm2020/lib/utils/obj-type.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
export class ObjTypePipe {
    transform(observations, type) {
        /* TODO: check layerlist for purity console.log("reevaluated"); */
        return observations.filter((o) => o.type === type);
    }
}
ObjTypePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ObjTypePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
ObjTypePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ObjTypePipe, name: "type", pure: false });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ObjTypePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'type', pure: false
                }]
        }] });
/**
 * e.g. usage *ngFor="let item of items | itemsfilter: callbackfn"
 */
export class ItemsFilterPipe {
    transform(items, callbackfn) {
        /* TODO: check layerlist for purity console.log("reevaluated"); */
        return items.filter(callbackfn);
    }
}
ItemsFilterPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ItemsFilterPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
ItemsFilterPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ItemsFilterPipe, name: "itemsfilter", pure: false });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ItemsFilterPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'itemsfilter', pure: false
                }]
        }] });
