// @dlr-eoc/services-map-state/esm2020/lib/map-state.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { MapState } from './types/map-state';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
const initialState = new MapState(0, { lat: 0, lon: 0 });
export class MapStateService {
    constructor() {
        this.mapState = new BehaviorSubject(initialState);
        this.lastAction = new BehaviorSubject(null);
    }
    getMapState() {
        return this.mapState;
    }
    setMapState(state) {
        if (!state) {
            return;
        }
        this.lastAction.next('setState');
        if (state instanceof MapState) {
            const newState = new MapState(state.zoom, state.center, state.options, state.extent);
            this.mapState.next(newState);
        }
        else {
            const stateOptions = { ...{ notifier: 'user' }, ...state.options };
            const newState = new MapState(state.zoom, state.center, stateOptions, state.extent);
            this.mapState.next(newState);
        }
    }
    getExtent() {
        return this.mapState.pipe(map((state) => state.extent));
    }
    setExtent(extent, notifier = 'user') {
        if (!extent) {
            return;
        }
        this.lastAction.next('setExtent');
        const state = this.getMapState().getValue();
        state.options.notifier = notifier;
        const newState = new MapState(state.zoom, state.center, state.options, extent);
        this.mapState.next(newState);
    }
    getLastAction() {
        return this.lastAction;
    }
}
MapStateService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: MapStateService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
MapStateService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: MapStateService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: MapStateService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return []; } });
