// @dlr-eoc/services-layers/esm2020/lib/layers.module.mjs
import { NgModule } from '@angular/core';
import { LayersService } from './layers.service';
import * as i0 from "@angular/core";
export class LayersModule {
}
LayersModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: LayersModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LayersModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: LayersModule });
LayersModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: LayersModule, providers: [LayersService], imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: LayersModule, decorators: [{
            type: NgModule,
            args: [{
                    // components and directives of the module
                    declarations: [],
                    // which other modules are used
                    imports: [],
                    // components and directives of the module
                    exports: [],
                    // Most of the time, these are services that you create and provide
                    providers: [LayersService]
                }]
        }] });
