// @dlr-eoc/layer-control/esm2020/lib/base-layer-control/base-layer-control.component.mjs
import { Component, Input } from '@angular/core';
import { LayerGroup } from '@dlr-eoc/services-layers';
import * as i0 from "@angular/core";
import * as i1 from "../layerentry-group/layerentry-group.component";
import * as i2 from "../layerentry/layerentry.component";
import * as i3 from "@angular/common";
import * as i4 from "../utils/array-reverse.pipe";
export class BaseLayerControlComponent {
    constructor() {
        this.layergroups = [];
    }
    ngOnInit() {
        this.layerGroupsSubscription = this.layersSvc.getLayerGroups().subscribe(layergroups => {
            this.layergroups = layergroups.filter((group) => group.filtertype === 'Baselayers');
            // console.log(this.layergroups)
        });
    }
    isLayerGroup(group) {
        if (group instanceof LayerGroup) {
            return true;
        }
        else {
            return false;
        }
    }
    ngOnDestroy() {
        this.layerGroupsSubscription.unsubscribe();
    }
}
BaseLayerControlComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: BaseLayerControlComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
BaseLayerControlComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.6", type: BaseLayerControlComponent, selector: "ukis-base-layer-control", inputs: { layersSvc: "layersSvc", mapStateSvc: "mapStateSvc" }, ngImport: i0, template: "<div class=\"list-item\" *ngFor=\"let group of layergroups | reverse;\">\n  <ukis-layerentry-group *ngIf=\"isLayerGroup(group)\" [layersSvc]=\"layersSvc\" [mapState]=\"mapStateSvc\" [group]=\"group\"\n    [layerGroups]=\"layergroups\"></ukis-layerentry-group>\n\n  <ukis-layerentry *ngIf=\"!isLayerGroup(group)\" [layerGroups]=\"layergroups\" [expanded]=\"false\" [layersSvc]=\"layersSvc\"\n    [mapState]=\"mapStateSvc\" [layer]=\"group\" [group]=\"null\"></ukis-layerentry>\n</div>\n", styles: [""], components: [{ type: i1.LayerentryGroupComponent, selector: "ukis-layerentry-group", inputs: ["layersSvc", "mapState", "group", "layerGroups"] }, { type: i2.LayerentryComponent, selector: "ukis-layerentry", inputs: ["layersSvc", "mapState", "layer", "group", "layerGroups", "expanded", "expandable"], outputs: ["update"] }], directives: [{ type: i3.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], pipes: { "reverse": i4.ReversePipe } });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: BaseLayerControlComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ukis-base-layer-control', template: "<div class=\"list-item\" *ngFor=\"let group of layergroups | reverse;\">\n  <ukis-layerentry-group *ngIf=\"isLayerGroup(group)\" [layersSvc]=\"layersSvc\" [mapState]=\"mapStateSvc\" [group]=\"group\"\n    [layerGroups]=\"layergroups\"></ukis-layerentry-group>\n\n  <ukis-layerentry *ngIf=\"!isLayerGroup(group)\" [layerGroups]=\"layergroups\" [expanded]=\"false\" [layersSvc]=\"layersSvc\"\n    [mapState]=\"mapStateSvc\" [layer]=\"group\" [group]=\"null\"></ukis-layerentry>\n</div>\n", styles: [""] }]
        }], ctorParameters: function () { return []; }, propDecorators: { layersSvc: [{
                type: Input,
                args: ['layersSvc']
            }], mapStateSvc: [{
                type: Input,
                args: ['mapStateSvc']
            }] } });
