// @clr/angular/esm2020/utils/for-type-ahead/for-type-ahead.mjs
/*
 * Copyright (c) 2016-2022 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, Input } from '@angular/core';
import { ForTypeAheadProvider } from './for-type-ahead.service';
import * as i0 from "@angular/core";
import * as i1 from "./for-type-ahead.service";
// The purpose of this directive is to provide the text content of its host element
// or text content that's passed to its input.
// This directive must stay optional. So do not embed this directive in a component module by default.
// If users need to use this directive, they should import its module on their side.
export class ClrForTypeAhead {
    constructor(el, forTypeAheadProvider) {
        this.el = el;
        this.forTypeAheadProvider = forTypeAheadProvider;
    }
    trimAndLowerCase(value) {
        return value.toLocaleLowerCase().trim();
    }
    get content() {
        return this._content;
    }
    set content(value) {
        if (value) {
            this._content = this.trimAndLowerCase(value);
            this.forTypeAheadProvider.textContent = this._content;
        }
        else {
            this._content = this.trimAndLowerCase(this.el.nativeElement.textContent);
            this.forTypeAheadProvider.textContent = this._content;
        }
    }
    ngOnInit() {
        if (!this._content) {
            this._content = this.trimAndLowerCase(this.el.nativeElement.textContent);
            this.forTypeAheadProvider.textContent = this._content;
        }
    }
}
ClrForTypeAhead.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ClrForTypeAhead, deps: [{ token: i0.ElementRef }, { token: i1.ForTypeAheadProvider }], target: i0.ɵɵFactoryTarget.Directive });
ClrForTypeAhead.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.0.3", type: ClrForTypeAhead, selector: "[clrForTypeAhead]", inputs: { content: ["clrForTypeAhead", "content"] }, providers: [ForTypeAheadProvider], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ClrForTypeAhead, decorators: [{
            type: Directive,
            args: [{
                    providers: [ForTypeAheadProvider],
                    selector: '[clrForTypeAhead]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.ForTypeAheadProvider }]; }, propDecorators: { content: [{
                type: Input,
                args: ['clrForTypeAhead']
            }] } });
