// @dlr-eoc/map-ol/esm2020/lib/map-ol.module.mjs
import { NgModule } from '@angular/core';
import { MapOlComponent } from './map-ol.component';
import { MapOlService } from './map-ol.service';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class MapOlModule {
}
MapOlModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: MapOlModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MapOlModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: MapOlModule, declarations: [MapOlComponent], imports: [CommonModule], exports: [MapOlComponent] });
MapOlModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: MapOlModule, providers: [MapOlService], imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: MapOlModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MapOlComponent],
                    imports: [
                        CommonModule
                    ],
                    exports: [MapOlComponent],
                    providers: [MapOlService]
                }]
        }] });
