// @dlr-eoc/map-ol/esm2020/lib/map-ol.component.mjs
import { Component, ViewEncapsulation, Input, ViewChild } from '@angular/core';
import { MapState } from '@dlr-eoc/services-map-state';
import { skip } from 'rxjs/operators';
import { WmtsLayertype, WmsLayertype, CustomLayer, GeojsonLayertype, WfsLayertype } from '@dlr-eoc/services-layers';
import { getUid as olGetUid } from 'ol/util';
import olLayer from 'ol/layer/Layer';
import olLayerGroup from 'ol/layer/Group';
import Attribution from 'ol/control/Attribution';
import ScaleLine from 'ol/control/ScaleLine';
import Zoom from 'ol/control/Zoom';
import olMousePosition from 'ol/control/MousePosition';
import { toStringXY } from 'ol/coordinate';
import olFullScreen from 'ol/control/FullScreen';
import olOverviewMap from 'ol/control/OverviewMap';
import olTileLayer from 'ol/layer/Tile';
import olOSM from 'ol/source/OSM';
import olRotate from 'ol/control/Rotate';
import olRasterSource from 'ol/source/Raster';
import olSourceCluster from 'ol/source/Cluster';
import * as i0 from "@angular/core";
import * as i1 from "./map-ol.service";
import * as i2 from "@angular/common";
const ID_KEY = 'id';
export class MapOlComponent {
    constructor(mapSvc, ngZone) {
        this.mapSvc = mapSvc;
        this.ngZone = ngZone;
        this.subs = [];
        this.privMapwidth = 0;
        this.removePopupsOnMouseLeave = (evt) => {
            this.mapSvc.removeAllPopups((item) => {
                return item.get('addEvent') === 'pointermove';
            });
        };
    }
    get mapwidth() {
        return this.privMapwidth;
    }
    set mapwidth(width) {
        this.privMapwidth = width;
        this.map.updateSize();
    }
    /**
     * - subscribe to layers oninit so they get pulled after view init
     */
    ngOnInit() {
        /** Subscribe to mapStateSvc before map is created */
        this.subscribeToMapState();
        this.initMap();
        this.subscribeToLayers();
    }
    /**
     * - set target of ol map after angular has rendered the element
     * - then subscribe to map events
     */
    ngAfterViewInit() {
        this.map.setTarget(this.mapDivView.nativeElement);
        /** Get last state from mapStateSvc and set it, so a User can set the initial MapState in a component on ngOnInit */
        const oldMapState = this.mapStateSvc.getMapState().getValue();
        this.setMapState(oldMapState);
        /** Subscribe to map events when the map completely created  */
        this.subscribeToMapEvents();
        this.map.getTargetElement().addEventListener('mouseleave', this.removePopupsOnMouseLeave);
    }
    ngAfterViewChecked() {
        /**
         * - compare map size to update Map Size on container resize
         * - set Timeout to also resize map on route change
         */
        if (this.mapDivView) {
            const mapWidth = this.mapDivView.nativeElement.offsetWidth;
            if (mapWidth !== this.mapwidth) {
                this.ngZone.runOutsideAngular(() => {
                    setTimeout(() => {
                        this.mapwidth = mapWidth;
                    }, 100);
                });
            }
        }
    }
    ngOnDestroy() {
        this.subs.forEach(sub => sub.unsubscribe());
        if (this.map) {
            this.map.un('moveend', this.mapOnMoveend);
            this.map.un('click', this.mapOnClickMove);
            this.map.un('dblclick', this.mapOnDclick);
            this.map.getTargetElement().removeEventListener('mouseleave', this.removePopupsOnMouseLeave);
            this.map.getInteractions().forEach((i) => {
                this.map.removeInteraction(i);
            });
        }
    }
    addUpdateLayers(layers, filtertype, layersunderneath) {
        /** get all underneath layers for zIndex */
        let otherlayerslength = 0;
        layersunderneath.forEach(itemType => {
            otherlayerslength += this.mapSvc.getLayers(itemType).length;
        });
        /** if length of layers has changed add new layers */
        if (layers.length !== this.mapSvc.getLayers(filtertype).length) {
            this.mapSvc.setUkisLayers(layers, filtertype);
            // if layers underneath add thhen to the zIndex of layer
            if (otherlayerslength > 0) {
                for (const layer of layers) {
                    const ollayer = this.mapSvc.getLayerByKey({ key: ID_KEY, value: layer.id }, filtertype);
                    if (ollayer) {
                        if (ollayer.getZIndex() !== layers.indexOf(layer) + otherlayerslength) {
                            ollayer.setZIndex(layers.indexOf(layer) + otherlayerslength);
                            this.setZIndexForGroup(ollayer, layers, layer, otherlayerslength);
                        }
                    }
                }
            }
        }
        else {
            /** if layers already on the map -length not changed- update them */
            for (const layer of layers) {
                const ollayer = this.mapSvc.getLayerByKey({ key: ID_KEY, value: layer.id }, filtertype);
                if (ollayer) {
                    if (ollayer.getVisible() !== layer.visible) {
                        ollayer.setVisible(layer.visible);
                    }
                    if (ollayer.getOpacity() !== layer.opacity) {
                        ollayer.setOpacity(layer.opacity);
                    }
                    if (layer instanceof CustomLayer && ollayer instanceof olLayer) {
                        const newSource = layer.custom_layer.getSource();
                        const oldSource = ollayer.getSource();
                        if (newSource && olGetUid(oldSource) !== olGetUid(newSource)) {
                            ollayer.setSource(newSource);
                            // https://github.com/dlr-eoc/ukis-frontend-libraries/issues/100
                            if (oldSource instanceof olRasterSource) {
                                oldSource.dispose();
                            }
                        }
                    }
                    else if (layer instanceof CustomLayer && layer.custom_layer instanceof olLayerGroup && ollayer instanceof olLayerGroup) {
                        const newLayers = layer.custom_layer.getLayers().getArray();
                        const oldLayers = ollayer.getLayers().getArray();
                        /** assume the order and length of layers is not changing and no more grouping!!! */
                        oldLayers.forEach((l, i) => {
                            const newLayer = newLayers[i];
                            if (l instanceof olLayer && newLayer instanceof olLayer) {
                                const oldSource = l.getSource();
                                const newSource = newLayer.getSource();
                                if (newSource && olGetUid(oldSource) !== olGetUid(newSource)) {
                                    l.setSource(newSource);
                                }
                            }
                        });
                    }
                    if (otherlayerslength > 0) {
                        if (ollayer.getZIndex() !== layers.indexOf(layer) + otherlayerslength) {
                            ollayer.setZIndex(layers.indexOf(layer) + otherlayerslength);
                            this.setZIndexForGroup(ollayer, layers, layer, otherlayerslength);
                        }
                    }
                    else {
                        if (ollayer.getZIndex() !== layers.indexOf(layer)) {
                            ollayer.setZIndex(layers.indexOf(layer));
                            this.setZIndexForGroup(ollayer, layers, layer);
                        }
                    }
                    this.updateLayerParamsWith(ollayer, layer);
                }
            }
        }
    }
    /** addresses an issue in openlayers: https://github.com/openlayers/openlayers/issues/6654 */
    setZIndexForGroup(ollayer, layers, layer, otherlayerslength) {
        if (ollayer instanceof olLayerGroup) {
            ollayer.getLayers().forEach(l => {
                if (otherlayerslength) {
                    l.setZIndex(layers.indexOf(layer) + otherlayerslength);
                }
                else {
                    l.setZIndex(layers.indexOf(layer));
                }
            });
        }
    }
    updateLayerParamsWith(oldLayer, newLayer) {
        switch (newLayer.type) {
            case WmsLayertype:
                this.updateWmsLayerParamsWith(oldLayer, newLayer);
                break;
            case WmtsLayertype:
                this.updateWmtsLayerParamsWith(oldLayer, newLayer);
                break;
            case GeojsonLayertype:
                this.updateGeojsonLayerParamsWith(oldLayer, newLayer);
                break;
            case WfsLayertype:
                this.updateWfsLayerParamsWith(oldLayer, newLayer);
                break;
            default:
                break;
        }
    }
    updateWfsLayerParamsWith(oldLayer, newLayer) {
        // step 1: update style
        if (newLayer.options && newLayer.options.style !== oldLayer.getStyle()) {
            oldLayer.setStyle(newLayer.options.style);
        }
        // step 2: update source
        if (oldLayer.getSource().getUrl() !== newLayer.url) {
            oldLayer.getSource().setUrl(newLayer.url);
        }
    }
    /**
     * TODO: set all other props of GeoJsonLayer.options and GeoJsonLayer.cluster (see: IVectorLayerOptions)
     */
    updateGeojsonLayerParamsWith(oldLayer, newGeojsonLayer) {
        const oldSource = oldLayer.getSource();
        if (oldSource) {
            if (newGeojsonLayer.data) {
                const features = this.mapSvc.geoJsonToFeatures(newGeojsonLayer.data);
                if (oldSource instanceof olSourceCluster) {
                    const vectorSource = oldSource.getSource();
                    vectorSource.clear();
                    vectorSource.addFeatures(features);
                }
                else {
                    oldSource.clear();
                    oldSource.addFeatures(features);
                }
            }
            else if (newGeojsonLayer.url) {
                if (oldSource instanceof olSourceCluster) {
                    const vectorSource = oldSource.getSource();
                    vectorSource.setUrl(newGeojsonLayer.url);
                }
                else {
                    oldSource.setUrl(newGeojsonLayer.url);
                }
            }
            // 'distance' in also checks for 0 values
            if (typeof newGeojsonLayer?.cluster === 'object' && 'distance' in newGeojsonLayer.cluster && oldSource instanceof olSourceCluster) {
                oldSource.setDistance(newGeojsonLayer.cluster.distance);
            }
        }
        if (newGeojsonLayer?.options?.style) {
            oldLayer.setStyle(newGeojsonLayer.options.style);
        }
    }
    updateWmsLayerParamsWith(oldLayer, newWmsLayer) {
        if (oldLayer instanceof olLayer) {
            const source = oldLayer.getSource();
            const oldParams = source.getParams();
            const newParams = newWmsLayer.params;
            if (!this.shallowEqual(oldParams, newParams)) {
                oldLayer.getSource().updateParams(newParams);
            }
        }
    }
    updateWmtsLayerParamsWith(oldLayer, newWmtsLayer) {
        // contrary to a wms-source, a wmts-source has neither 'getParams' nor 'updateParams', so we need to do this manually.
        if (oldLayer instanceof olLayer) {
            const source = oldLayer.getSource();
            const oldStyle = source.getStyle();
            const oldFormat = source.getFormat();
            const oldVersion = source.getVersion();
            const oldMatrix = source.getMatrixSet();
            const newStyle = newWmtsLayer.params.style;
            const newFormat = newWmtsLayer.params.format;
            const newVersion = newWmtsLayer.params.version;
            const newMatrix = newWmtsLayer.params.matrixSetOptions.matrixSet;
            if (newStyle !== undefined && oldStyle !== newStyle
                || newFormat !== undefined && oldFormat !== newFormat
                || newVersion !== undefined && oldVersion !== newVersion
                || newMatrix !== undefined && oldMatrix !== newMatrix) {
                // console.log(oldStyle, oldFormat, oldVersion, oldMatrix)
                // console.log(newStyle, newFormat, newVersion, newMatrix)
                const olFiltertype = newWmtsLayer.filtertype.toLowerCase();
                // this.mapSvc.setUkisLayer(newWmtsLayer, olFiltertype);
                this.mapSvc.updateUkisLayer(newWmtsLayer, olFiltertype);
            }
        }
    }
    shallowEqual(a, b) {
        // Create arrays of property names
        const aProps = Object.getOwnPropertyNames(a);
        const bProps = Object.getOwnPropertyNames(b);
        // If number of properties is different,
        // objects are not equivalent
        if (aProps.length !== bProps.length) {
            return false;
        }
        for (const propName of aProps) {
            // If values of same property are not equal,
            // objects are not equivalent
            if (a[propName] !== b[propName]) {
                return false;
            }
        }
        // If we made it this far, objects
        // are considered equivalent
        return true;
    }
    addUpdateBaseLayers(layers) {
        /** if length of layers has changed add new layers */
        if (layers.length !== this.mapSvc.getLayers('baselayers').length) {
            // set only one visible at start
            const visiblelayers = layers.filter(l => l.visible === true);
            if (visiblelayers.length === 0) {
                layers[0].visible = true;
            }
            else if (visiblelayers.length > 1) {
                layers.forEach(l => l.visible = false);
                layers[0].visible = true;
            }
            this.mapSvc.setUkisLayers(layers, 'baselayers');
        }
        else {
            /** if layers already on the map -length not changed- update them */
            for (const layer of layers) {
                const bllayer = this.mapSvc.getLayerByKey({ key: ID_KEY, value: layer.id }, 'baselayers');
                if (bllayer) {
                    if (bllayer.getVisible() !== layer.visible) {
                        bllayer.setVisible(layer.visible);
                    }
                    if (bllayer.getZIndex() !== layers.indexOf(layer)) {
                        bllayer.setZIndex(layers.indexOf(layer));
                    }
                    if (bllayer.getOpacity() !== layer.opacity) {
                        bllayer.setOpacity(layer.opacity);
                    }
                }
            }
        }
    }
    subscribeToLayers() {
        // ----------------------
        // add/remove layers
        if (this.layersSvc) {
            const baselayersAddRemoveOn = this.layersSvc.getBaseLayers().subscribe(layers => this.addUpdateBaseLayers(layers));
            this.subs.push(baselayersAddRemoveOn);
            const onLayers = this.layersSvc.getLayers().subscribe(layers => this.addUpdateLayers(layers, 'layers', ['baselayers']));
            this.subs.push(onLayers);
            const onOverlays = this.layersSvc.getOverlays().subscribe(layers => this.addUpdateLayers(layers, 'overlays', ['baselayers', 'layers']));
            this.subs.push(onOverlays);
        }
    }
    setMapState(mapState) {
        const lastAction = this.mapStateSvc.getLastAction().getValue();
        if (mapState.options.notifier === 'user') {
            if (lastAction === 'setExtent') {
                this.mapSvc.setExtent(mapState.extent, true);
            }
            else if (lastAction === 'setState') {
                this.mapSvc.setZoom(mapState.zoom, mapState.options.notifier);
                this.mapSvc.setCenter([mapState.center.lon, mapState.center.lat], true);
            }
        }
        /* else if (mapState.options.notifier === 'map') {
          console.log("--------Map triggered mapState change", mapState);
        } */
    }
    subscribeToMapState() {
        if (this.mapStateSvc) {
            /** .pipe(skip(1)) skips the first, e.g. initial value of the BehaviorSubject!! -- https://www.learnrxjs.io/learn-rxjs/operators/filtering/skip#why-use-skip  */
            const mapStateOn = this.mapStateSvc.getMapState().pipe(skip(1)).subscribe(state => this.setMapState(state));
            this.subs.push(mapStateOn);
        }
    }
    subscribeToMapEvents() {
        this.mapOnMoveend = (evt) => {
            // const zoom = Math.round(this.mapSvc.getZoom());
            const zoom = this.mapSvc.getZoom();
            const center = this.mapSvc.getCenter(true);
            const extent = this.mapSvc.getCurrentExtent(true);
            // const newCenter = { lat: parseFloat(center[1].toFixed(6)), lon: parseFloat(center[0].toFixed(6)) };
            const newCenter = { lat: parseFloat(center[1]), lon: parseFloat(center[0]) };
            const ms = new MapState(zoom, newCenter, { notifier: 'map' }, extent);
            this.mapStateSvc.setMapState(ms);
        };
        this.map.on('moveend', this.mapOnMoveend);
        /** handle click and pointermove/mousemove */
        this.mapOnClickMove = (evt) => {
            this.mapSvc.layersOnMapEvent(evt);
        };
        this.map.on(['click', 'pointermove'], this.mapOnClickMove);
        /** handle double click */
        this.mapOnDclick = (evt) => {
            this.mapSvc.removeAllPopups();
        };
        this.map.on('dblclick', this.mapOnDclick);
    }
    initMap() {
        const olMapView = this.mapSvc.createMap();
        this.map = olMapView.map; //
        this.setControls();
        if (!this.layersSvc) {
            console.log('there is no layersSvc as defined!');
        }
        if (!this.mapStateSvc) {
            console.log('there is no mapStateSvc as defined!');
        }
    }
    setControls() {
        // add Control only if this functions is defined
        const tempControls = [];
        const oldControls = [];
        if (this.map) {
            const controlsArry = this.map.getControls().getArray();
            controlsArry.forEach(i => oldControls.push(i));
            /** fix: The Attribution Control is displayed twice #3 */
            this.map.getControls().clear();
        }
        if (this.controls && this.map) {
            if (this.controls.attribution !== false) {
                let attributionOptions = {
                    collapsible: true,
                    collapsed: false
                };
                if (typeof this.controls.attribution === 'object') {
                    attributionOptions = Object.assign(attributionOptions, this.controls.attribution);
                }
                const attribution = new Attribution(attributionOptions);
                tempControls.push(attribution);
            }
            if (this.controls.scaleLine) {
                let scaleLineOptions = {};
                if (typeof this.controls.scaleLine === 'object') {
                    scaleLineOptions = Object.assign(scaleLineOptions, this.controls.scaleLine);
                }
                const scaleLineControl = new ScaleLine(scaleLineOptions);
                tempControls.push(scaleLineControl);
            }
            if (this.controls.zoom !== false) {
                let zoomOptions = {};
                if (typeof this.controls.zoom === 'object') {
                    zoomOptions = Object.assign(zoomOptions, this.controls.zoom);
                }
                const zoomControl = new Zoom(zoomOptions);
                tempControls.push(zoomControl);
            }
            if (this.controls.mousePosition) {
                let mousePositionOptions = {
                    coordinateFormat: coordinate => {
                        return toStringXY(coordinate, 2);
                    },
                    projection: 'EPSG:4326'
                };
                if (typeof this.controls.mousePosition === 'object') {
                    mousePositionOptions = Object.assign(mousePositionOptions, this.controls.mousePosition);
                }
                const mousePosition = new olMousePosition(mousePositionOptions);
                tempControls.push(mousePosition);
            }
            if (this.controls.fullScreen) {
                let fullScreenOptions = {};
                if (typeof this.controls.fullScreen === 'object') {
                    fullScreenOptions = Object.assign(fullScreenOptions, this.controls.fullScreen);
                }
                const fullScreen = new olFullScreen(fullScreenOptions);
                tempControls.push(fullScreen);
            }
            if (this.controls.overviewMap) {
                const icon = document.createElement('spawn');
                icon.innerHTML = '<clr-icon shape="world"></clr-icon>';
                let overviewMapOptions = {
                    layers: [new olTileLayer({
                            source: new olOSM()
                        })],
                    label: icon
                };
                if (typeof this.controls.overviewMap === 'object') {
                    overviewMapOptions = Object.assign(overviewMapOptions, this.controls.overviewMap);
                }
                const overviewMap = new olOverviewMap(overviewMapOptions);
                tempControls.push(overviewMap);
            }
            if (this.controls.rotate) {
                let rotateOptions = {};
                if (typeof this.controls.rotate === 'object') {
                    rotateOptions = Object.assign(rotateOptions, this.controls.rotate);
                }
                const rotate = new olRotate(rotateOptions);
                tempControls.push(rotate);
            }
            if (tempControls.length) {
                /**
                 * check tempControls dose not include oldControls
                 * https://medium.com/@alvaro.saburido/set-theory-for-arrays-in-es6-eb2f20a61848#f22b
                 */
                const difference = oldControls.filter(x => !tempControls.includes(x));
                difference.forEach(i => tempControls.push(i));
                this.map.getControls().extend(tempControls);
            }
        }
    }
}
MapOlComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: MapOlComponent, deps: [{ token: i1.MapOlService }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Component });
MapOlComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.6", type: MapOlComponent, selector: "ukis-map-ol", inputs: { layersSvc: "layersSvc", mapStateSvc: ["mapState", "mapStateSvc"], controls: "controls" }, viewQueries: [{ propertyName: "mapDivView", first: true, predicate: ["mapDiv"], descendants: true }], ngImport: i0, template: "<div #mapDiv id=\"map\" ref=\"map\" class=\"map\">\n  <div *ngIf=\"controls && controls.crosshair\" id=\"crosshair\" class=\"crosshair\">+</div>\n</div>\n", styles: ["@charset \"UTF-8\";.ol-box{box-sizing:border-box;border-radius:2px;border:1.5px solid rgb(179,197,219);background-color:#fff6}.ol-mouse-position{top:8px;right:8px;position:absolute}.ol-scale-line{background:rgba(0,60,136,.3);border-radius:4px;bottom:8px;left:8px;padding:2px;position:absolute}.ol-scale-line-inner{border:1px solid #eee;border-top:none;color:#eee;font-size:10px;text-align:center;margin:1px;will-change:contents,width;transition:all .25s}.ol-scale-bar{position:absolute;bottom:8px;left:8px}.ol-scale-step-marker{width:1px;height:15px;background-color:#000;float:right;z-index:10}.ol-scale-step-text{position:absolute;bottom:-5px;font-size:12px;z-index:11;color:#000;text-shadow:-2px 0 #FFFFFF,0 2px #FFFFFF,2px 0 #FFFFFF,0 -2px #FFFFFF}.ol-scale-text{position:absolute;font-size:14px;text-align:center;bottom:25px;color:#000;text-shadow:-2px 0 #FFFFFF,0 2px #FFFFFF,2px 0 #FFFFFF,0 -2px #FFFFFF}.ol-scale-singlebar{position:relative;height:10px;z-index:9;box-sizing:border-box;border:1px solid black}.ol-unsupported{display:none}.ol-viewport,.ol-unselectable{-webkit-touch-callout:none;-webkit-user-select:none;user-select:none;-webkit-tap-highlight-color:rgba(0,0,0,0)}.ol-viewport canvas{all:unset}.ol-selectable{-webkit-touch-callout:default;-webkit-user-select:text;user-select:text}.ol-grabbing{cursor:grabbing}.ol-grab{cursor:move;cursor:grab}.ol-control{position:absolute;background-color:#fff6;border-radius:4px;padding:2px}.ol-control:hover{background-color:#fff9}.ol-zoom{top:.5em;left:.5em}.ol-rotate{top:.5em;right:.5em;transition:opacity .25s linear,visibility 0s linear}.ol-rotate.ol-hidden{opacity:0;visibility:hidden;transition:opacity .25s linear,visibility 0s linear .25s}.ol-zoom-extent{top:4.643em;left:.5em}.ol-full-screen{right:.5em;top:.5em}.ol-control button{display:block;margin:1px;padding:0;color:#fff;font-weight:700;text-decoration:none;font-size:inherit;text-align:center;height:1.375em;width:1.375em;line-height:.4em;background-color:#003c8880;border:none;border-radius:2px}.ol-control button::-moz-focus-inner{border:none;padding:0}.ol-zoom-extent button{line-height:1.4em}.ol-compass{display:block;font-weight:400;font-size:1.2em;will-change:transform}.ol-touch .ol-control button{font-size:1.5em}.ol-touch .ol-zoom-extent{top:5.5em}.ol-control button:hover,.ol-control button:focus{text-decoration:none;background-color:#003c88b3}.ol-zoom .ol-zoom-in{border-radius:2px 2px 0 0}.ol-zoom .ol-zoom-out{border-radius:0 0 2px 2px}.ol-attribution{text-align:right;bottom:.5em;right:.5em;max-width:calc(100% - 1.3em);display:flex;flex-flow:row-reverse;align-items:center}.ol-attribution a{color:#003c88b3;text-decoration:none}.ol-attribution ul{margin:0;padding:1px .5em;color:#000;text-shadow:0 0 2px #fff;font-size:12px}.ol-attribution li{display:inline;list-style:none}.ol-attribution li:not(:last-child):after{content:\" \"}.ol-attribution img{max-height:2em;max-width:inherit;vertical-align:middle}.ol-attribution button{flex-shrink:0}.ol-attribution.ol-collapsed ul{display:none}.ol-attribution:not(.ol-collapsed){background:rgba(255,255,255,.8)}.ol-attribution.ol-uncollapsible{bottom:0;right:0;border-radius:4px 0 0}.ol-attribution.ol-uncollapsible img{margin-top:-.2em;max-height:1.6em}.ol-attribution.ol-uncollapsible button{display:none}.ol-zoomslider{top:4.5em;left:.5em;height:200px}.ol-zoomslider button{position:relative;height:10px}.ol-touch .ol-zoomslider{top:5.5em}.ol-overviewmap{left:.5em;bottom:.5em}.ol-overviewmap.ol-uncollapsible{bottom:0;left:0;border-radius:0 4px 0 0}.ol-overviewmap .ol-overviewmap-map,.ol-overviewmap button{display:block}.ol-overviewmap .ol-overviewmap-map{border:1px solid #7b98bc;height:150px;margin:2px;width:150px}.ol-overviewmap:not(.ol-collapsed) button{bottom:2px;left:2px;position:absolute}.ol-overviewmap.ol-collapsed .ol-overviewmap-map,.ol-overviewmap.ol-uncollapsible button{display:none}.ol-overviewmap:not(.ol-collapsed){background:rgba(255,255,255,.8)}.ol-overviewmap-box{border:2px dotted rgba(0,60,136,.7)}.ol-overviewmap .ol-overviewmap-box:hover{cursor:move}:root{--ukis-crosshair-color: #ff0004;--ukis-popup-bg-color: rgb(238, 238, 238);--ukis-drop-shadow: drop-shadow(0 1px 4px rgba(0, 0, 0, .2));--ukis-ol-bbox-bg-color: rgba(255, 255, 255, .4);--ukis-ol-bbox-border-color: rgba(87, 87, 87, .4);--ukis-ol-overviewmap-left: .5em;--ukis-ol-overviewmap-bottom: 3em;--ukis-ol-control-bg-color: rgba(87, 87, 87, .6);--ukis-ol-control-border-color: rgba(87, 87, 87, .4)}.map{width:100%;height:100%;position:relative}.ol-popup{position:absolute;background-color:var(--ukis-popup-bg-color);filter:var(--ukis-drop-shadow);padding:15px;border-radius:2px;bottom:12px;left:-50px;min-width:280px;display:none;z-index:1000000}.ol-popup .ol-popup-closer{text-decoration:none;position:absolute;top:0px;right:8px;cursor:pointer}.ol-popup .ol-popup-closer:after{content:\"\\2716\"}.ol-popup:after,.ol-popup:before{top:100%;border:solid transparent;content:\" \";height:0;width:0;position:absolute;pointer-events:none}.ol-popup:after{border-top-color:var(--ukis-popup-bg-color);border-width:10px;left:48px;margin-left:-10px}.ol-popup:before{border-width:11px;left:48px;margin-left:-11px}.crosshair{top:50%;left:50%;font-size:large;color:var(--ukis-crosshair-color);z-index:1;position:absolute;cursor:default}.ol-control button{cursor:pointer;background-color:var(--ukis-ol-control-bg-color)}.ol-control button:focus{background-color:var(--ukis-ol-control-bg-color)}.ol-control button:hover{background-color:var(--ukis-ol-control-border-color)}.ol-scale-line{background:rgba(255,255,255,.4);line-height:1.375em}.ol-scale-line .ol-scale-line-inner{background:var(--ukis-ol-control-bg-color);border-color:var(--ukis-ol-control-bg-color)}.ol-attribution ul{font-size:.5rem}.ol-box{background-color:var(--ukis-ol-bbox-bg-color);border-color:var(--ukis-ol-bbox-border-color)!important}.ol-box.ol-drag-select{border-color:var(--clr-color-action-600, var(--fallback-color-no-clr))!important}.ol-overviewmap{left:var(--ukis-ol-overviewmap-left)!important;bottom:var(--ukis-ol-overviewmap-bottom)!important}\n"], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: MapOlComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ukis-map-ol', encapsulation: ViewEncapsulation.None, template: "<div #mapDiv id=\"map\" ref=\"map\" class=\"map\">\n  <div *ngIf=\"controls && controls.crosshair\" id=\"crosshair\" class=\"crosshair\">+</div>\n</div>\n", styles: ["@charset \"UTF-8\";.ol-box{box-sizing:border-box;border-radius:2px;border:1.5px solid rgb(179,197,219);background-color:#fff6}.ol-mouse-position{top:8px;right:8px;position:absolute}.ol-scale-line{background:rgba(0,60,136,.3);border-radius:4px;bottom:8px;left:8px;padding:2px;position:absolute}.ol-scale-line-inner{border:1px solid #eee;border-top:none;color:#eee;font-size:10px;text-align:center;margin:1px;will-change:contents,width;transition:all .25s}.ol-scale-bar{position:absolute;bottom:8px;left:8px}.ol-scale-step-marker{width:1px;height:15px;background-color:#000;float:right;z-index:10}.ol-scale-step-text{position:absolute;bottom:-5px;font-size:12px;z-index:11;color:#000;text-shadow:-2px 0 #FFFFFF,0 2px #FFFFFF,2px 0 #FFFFFF,0 -2px #FFFFFF}.ol-scale-text{position:absolute;font-size:14px;text-align:center;bottom:25px;color:#000;text-shadow:-2px 0 #FFFFFF,0 2px #FFFFFF,2px 0 #FFFFFF,0 -2px #FFFFFF}.ol-scale-singlebar{position:relative;height:10px;z-index:9;box-sizing:border-box;border:1px solid black}.ol-unsupported{display:none}.ol-viewport,.ol-unselectable{-webkit-touch-callout:none;-webkit-user-select:none;user-select:none;-webkit-tap-highlight-color:rgba(0,0,0,0)}.ol-viewport canvas{all:unset}.ol-selectable{-webkit-touch-callout:default;-webkit-user-select:text;user-select:text}.ol-grabbing{cursor:grabbing}.ol-grab{cursor:move;cursor:grab}.ol-control{position:absolute;background-color:#fff6;border-radius:4px;padding:2px}.ol-control:hover{background-color:#fff9}.ol-zoom{top:.5em;left:.5em}.ol-rotate{top:.5em;right:.5em;transition:opacity .25s linear,visibility 0s linear}.ol-rotate.ol-hidden{opacity:0;visibility:hidden;transition:opacity .25s linear,visibility 0s linear .25s}.ol-zoom-extent{top:4.643em;left:.5em}.ol-full-screen{right:.5em;top:.5em}.ol-control button{display:block;margin:1px;padding:0;color:#fff;font-weight:700;text-decoration:none;font-size:inherit;text-align:center;height:1.375em;width:1.375em;line-height:.4em;background-color:#003c8880;border:none;border-radius:2px}.ol-control button::-moz-focus-inner{border:none;padding:0}.ol-zoom-extent button{line-height:1.4em}.ol-compass{display:block;font-weight:400;font-size:1.2em;will-change:transform}.ol-touch .ol-control button{font-size:1.5em}.ol-touch .ol-zoom-extent{top:5.5em}.ol-control button:hover,.ol-control button:focus{text-decoration:none;background-color:#003c88b3}.ol-zoom .ol-zoom-in{border-radius:2px 2px 0 0}.ol-zoom .ol-zoom-out{border-radius:0 0 2px 2px}.ol-attribution{text-align:right;bottom:.5em;right:.5em;max-width:calc(100% - 1.3em);display:flex;flex-flow:row-reverse;align-items:center}.ol-attribution a{color:#003c88b3;text-decoration:none}.ol-attribution ul{margin:0;padding:1px .5em;color:#000;text-shadow:0 0 2px #fff;font-size:12px}.ol-attribution li{display:inline;list-style:none}.ol-attribution li:not(:last-child):after{content:\" \"}.ol-attribution img{max-height:2em;max-width:inherit;vertical-align:middle}.ol-attribution button{flex-shrink:0}.ol-attribution.ol-collapsed ul{display:none}.ol-attribution:not(.ol-collapsed){background:rgba(255,255,255,.8)}.ol-attribution.ol-uncollapsible{bottom:0;right:0;border-radius:4px 0 0}.ol-attribution.ol-uncollapsible img{margin-top:-.2em;max-height:1.6em}.ol-attribution.ol-uncollapsible button{display:none}.ol-zoomslider{top:4.5em;left:.5em;height:200px}.ol-zoomslider button{position:relative;height:10px}.ol-touch .ol-zoomslider{top:5.5em}.ol-overviewmap{left:.5em;bottom:.5em}.ol-overviewmap.ol-uncollapsible{bottom:0;left:0;border-radius:0 4px 0 0}.ol-overviewmap .ol-overviewmap-map,.ol-overviewmap button{display:block}.ol-overviewmap .ol-overviewmap-map{border:1px solid #7b98bc;height:150px;margin:2px;width:150px}.ol-overviewmap:not(.ol-collapsed) button{bottom:2px;left:2px;position:absolute}.ol-overviewmap.ol-collapsed .ol-overviewmap-map,.ol-overviewmap.ol-uncollapsible button{display:none}.ol-overviewmap:not(.ol-collapsed){background:rgba(255,255,255,.8)}.ol-overviewmap-box{border:2px dotted rgba(0,60,136,.7)}.ol-overviewmap .ol-overviewmap-box:hover{cursor:move}:root{--ukis-crosshair-color: #ff0004;--ukis-popup-bg-color: rgb(238, 238, 238);--ukis-drop-shadow: drop-shadow(0 1px 4px rgba(0, 0, 0, .2));--ukis-ol-bbox-bg-color: rgba(255, 255, 255, .4);--ukis-ol-bbox-border-color: rgba(87, 87, 87, .4);--ukis-ol-overviewmap-left: .5em;--ukis-ol-overviewmap-bottom: 3em;--ukis-ol-control-bg-color: rgba(87, 87, 87, .6);--ukis-ol-control-border-color: rgba(87, 87, 87, .4)}.map{width:100%;height:100%;position:relative}.ol-popup{position:absolute;background-color:var(--ukis-popup-bg-color);filter:var(--ukis-drop-shadow);padding:15px;border-radius:2px;bottom:12px;left:-50px;min-width:280px;display:none;z-index:1000000}.ol-popup .ol-popup-closer{text-decoration:none;position:absolute;top:0px;right:8px;cursor:pointer}.ol-popup .ol-popup-closer:after{content:\"\\2716\"}.ol-popup:after,.ol-popup:before{top:100%;border:solid transparent;content:\" \";height:0;width:0;position:absolute;pointer-events:none}.ol-popup:after{border-top-color:var(--ukis-popup-bg-color);border-width:10px;left:48px;margin-left:-10px}.ol-popup:before{border-width:11px;left:48px;margin-left:-11px}.crosshair{top:50%;left:50%;font-size:large;color:var(--ukis-crosshair-color);z-index:1;position:absolute;cursor:default}.ol-control button{cursor:pointer;background-color:var(--ukis-ol-control-bg-color)}.ol-control button:focus{background-color:var(--ukis-ol-control-bg-color)}.ol-control button:hover{background-color:var(--ukis-ol-control-border-color)}.ol-scale-line{background:rgba(255,255,255,.4);line-height:1.375em}.ol-scale-line .ol-scale-line-inner{background:var(--ukis-ol-control-bg-color);border-color:var(--ukis-ol-control-bg-color)}.ol-attribution ul{font-size:.5rem}.ol-box{background-color:var(--ukis-ol-bbox-bg-color);border-color:var(--ukis-ol-bbox-border-color)!important}.ol-box.ol-drag-select{border-color:var(--clr-color-action-600, var(--fallback-color-no-clr))!important}.ol-overviewmap{left:var(--ukis-ol-overviewmap-left)!important;bottom:var(--ukis-ol-overviewmap-bottom)!important}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.MapOlService }, { type: i0.NgZone }]; }, propDecorators: { mapDivView: [{
                type: ViewChild,
                args: ['mapDiv']
            }], layersSvc: [{
                type: Input,
                args: ['layersSvc']
            }], mapStateSvc: [{
                type: Input,
                args: ['mapState']
            }], controls: [{
                type: Input,
                args: ['controls']
            }] } });
