// @clr/angular/esm2020/timeline/timeline.mjs
/*
 * Copyright (c) 2016-2022 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, Input, HostBinding } from '@angular/core';
import { ClrTimelineLayout } from './enums/timeline-layout.enum';
import { TimelineIconAttributeService } from './providers/timeline-icon-attribute.service';
import * as i0 from "@angular/core";
export class ClrTimeline {
    constructor() {
        this.layout = ClrTimelineLayout.HORIZONTAL;
    }
    get isVertical() {
        return this.layout === ClrTimelineLayout.VERTICAL;
    }
}
ClrTimeline.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ClrTimeline, deps: [], target: i0.ɵɵFactoryTarget.Component });
ClrTimeline.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.0.3", type: ClrTimeline, selector: "clr-timeline", inputs: { layout: ["clrLayout", "layout"] }, host: { properties: { "class.clr-timeline": "true", "class.clr-timeline-vertical": "this.isVertical" } }, providers: [TimelineIconAttributeService], ngImport: i0, template: ` <ng-content></ng-content> `, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ClrTimeline, decorators: [{
            type: Component,
            args: [{
                    selector: 'clr-timeline',
                    template: ` <ng-content></ng-content> `,
                    host: { '[class.clr-timeline]': 'true' },
                    providers: [TimelineIconAttributeService],
                }]
        }], propDecorators: { layout: [{
                type: Input,
                args: ['clrLayout']
            }], isVertical: [{
                type: HostBinding,
                args: ['class.clr-timeline-vertical']
            }] } });
