// @dlr-eoc/layer-control/esm2020/lib/layer-control.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ClarityModule } from '@clr/angular';
import { FormsModule } from '@angular/forms';
import { LayerControlComponent } from './layer-control/layer-control.component';
import { LayerentryComponent } from './layerentry/layerentry.component';
import { LayerentryGroupComponent } from './layerentry-group/layerentry-group.component';
import { ObjTypePipe, ItemsFilterPipe } from './utils/obj-type.pipe';
import { ReversePipe } from './utils/array-reverse.pipe';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { BaseLayerControlComponent } from './base-layer-control/base-layer-control.component';
import { CoreUiModule } from '@dlr-eoc/core-ui';
import * as i0 from "@angular/core";
export class LayerControlModule {
}
LayerControlModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: LayerControlModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LayerControlModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: LayerControlModule, declarations: [LayerControlComponent, LayerentryComponent, LayerentryGroupComponent, ObjTypePipe, ItemsFilterPipe, ReversePipe, BaseLayerControlComponent], imports: [CommonModule, ClarityModule, FormsModule, DragDropModule, CoreUiModule], exports: [LayerControlComponent, LayerentryComponent, LayerentryGroupComponent, ObjTypePipe, ItemsFilterPipe, ReversePipe, BaseLayerControlComponent] });
LayerControlModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: LayerControlModule, imports: [[CommonModule, ClarityModule, FormsModule, DragDropModule, CoreUiModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: LayerControlModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LayerControlComponent, LayerentryComponent, LayerentryGroupComponent, ObjTypePipe, ItemsFilterPipe, ReversePipe, BaseLayerControlComponent],
                    imports: [CommonModule, ClarityModule, FormsModule, DragDropModule, CoreUiModule],
                    exports: [LayerControlComponent, LayerentryComponent, LayerentryGroupComponent, ObjTypePipe, ItemsFilterPipe, ReversePipe, BaseLayerControlComponent],
                }]
        }] });
