// @dlr-eoc/utils-maps/esm2020/lib/webgl/engine.core.mjs
import { createShaderProgram, setup3dScene, createFloatBuffer, getAttributeLocation, bindBufferToAttribute, getUniformLocation, bindValueToUniform, clearBackground, bindProgram, createTexture, bindTextureToUniform, bindFramebuffer, bindOutputCanvasToFramebuffer, updateBufferData, bindTextureToFramebuffer, createEmptyTexture, createFramebuffer, updateTexture, createIndexBuffer, drawArray, drawElements, bindIndexBuffer, createDataTexture } from './webgl';
// dead-simple hash function - not intended to be secure in any way.
const hash = function (s) {
    let h = 0;
    for (const c of s) {
        h += c.charCodeAt(0);
    }
    return `${h}`;
};
export class Program {
    constructor(gl, vertexShaderSource, fragmentShaderSource) {
        this.vertexShaderSource = vertexShaderSource;
        this.fragmentShaderSource = fragmentShaderSource;
        this.program = createShaderProgram(gl, vertexShaderSource, fragmentShaderSource);
        this.id = hash(vertexShaderSource + fragmentShaderSource);
    }
}
export class Uniform {
    constructor(gl, program, variableName, type, data) {
        this.location = getUniformLocation(gl, program.program, variableName);
        this.type = type;
        this.value = data;
        this.variableName = variableName;
    }
}
export class Texture {
    constructor(gl, program, variableName, im, bindPoint) {
        this.location = getUniformLocation(gl, program.program, variableName);
        if (im instanceof HTMLImageElement || im instanceof HTMLCanvasElement) {
            this.texture = createTexture(gl, im);
        }
        else {
            this.texture = im;
        }
        this.bindPoint = bindPoint;
        this.variableName = variableName;
    }
}
export class DataTexture {
    constructor(gl, program, variableName, data, bindPoint) {
        this.location = getUniformLocation(gl, program.program, variableName);
        this.texture = createDataTexture(gl, data);
        this.bindPoint = bindPoint;
        this.variableName = variableName;
    }
}
export class Attribute {
    constructor(gl, program, variableName, data, drawingMode = 'triangles') {
        let glDrawingMode;
        switch (drawingMode) {
            case 'triangles':
                glDrawingMode = gl.TRIANGLES;
                break;
            case 'lines':
                glDrawingMode = gl.LINES;
                break;
            case 'points':
                glDrawingMode = gl.POINTS;
                break;
            default:
                throw new Error(`Invalid drawing mode ${drawingMode}`);
        }
        this.location = getAttributeLocation(gl, program.program, variableName);
        this.value = createFloatBuffer(gl, data, glDrawingMode);
        this.variableName = variableName;
        this.drawingMode = glDrawingMode;
    }
}
export class Index {
    constructor(gl, indices, drawingMode = 'triangles') {
        let glDrawingMode;
        switch (drawingMode) {
            case 'triangles':
                glDrawingMode = gl.TRIANGLES;
                break;
            case 'lines':
                glDrawingMode = gl.LINES;
                break;
            case 'points':
                glDrawingMode = gl.POINTS;
                break;
            default:
                throw new Error(`Invalid drawing mode ${drawingMode}`);
        }
        this.index = createIndexBuffer(gl, indices, glDrawingMode);
    }
}
function first(arr, condition) {
    for (const el of arr) {
        if (condition(el)) {
            return el;
        }
    }
    return null;
}
function parseProgram(program) {
    const attributeRegex = /^\s*attribute (int|float|vec2|vec3|vec4|mat2|mat3|mat4) (\w*);/gm;
    const uniformRegex = /^\s*uniform (int|float|vec2|vec3|vec4|mat2|mat3|mat4) (\w*)(\[\d\])*;/gm;
    const textureRegex = /^\s*uniform sampler2D (\w*);/gm;
    const precisionRegex = /^\s*precision (\w*) float;/gm;
    const shaderCode = program.fragmentShaderSource + '\n\n\n' + program.vertexShaderSource;
    const attributeNames = [];
    let attributeMatches;
    while ((attributeMatches = attributeRegex.exec(shaderCode)) !== null) {
        attributeNames.push(attributeMatches[2]);
    }
    const uniformNames = [];
    let uniformMatches;
    while ((uniformMatches = uniformRegex.exec(shaderCode)) !== null) {
        uniformNames.push(uniformMatches[2]);
    }
    const textureNames = [];
    let textureMatches;
    while ((textureMatches = textureRegex.exec(shaderCode)) !== null) {
        textureNames.push(textureMatches[1]);
    }
    const precisions = [];
    let precisionMatches;
    while ((precisionMatches = precisionRegex.exec(shaderCode)) !== null) {
        precisions.push(precisionMatches[1]);
    }
    return [attributeNames, uniformNames, textureNames, precisions];
}
export class Shader {
    constructor(program, attributes, uniforms, textures, index) {
        this.program = program;
        this.attributes = attributes;
        this.uniforms = uniforms;
        this.textures = textures;
        this.index = index;
        const [attributeNames, uniformNames, textureNames, precisions] = parseProgram(program);
        for (const attrName of attributeNames) {
            const found = attributes.filter(a => a.variableName === attrName);
            if (found.length !== 1) {
                throw new Error(`Provided ${found.length} values for shader's attribute ${attrName}.`);
            }
        }
        for (const uniformName of uniformNames) {
            const found = uniforms.filter(a => a.variableName === uniformName);
            if (found.length !== 1) {
                throw new Error(`Provided ${found.length} values for shader's uniform ${uniformName}.`);
            }
        }
        for (const texName of textureNames) {
            const found = textures.filter(a => a.variableName === texName);
            if (found.length !== 1) {
                throw new Error(`Provided ${found.length} values for shader's texture ${texName}.`);
            }
        }
        if (precisions.length === 1) {
            console.warn(`You have only provided one precision qualifier.
            This can cause issues when you want to use a uniform in both the vertex- and the fragment-shader.`);
        }
        const lengths = this.attributes.map(a => a.value.vectorCount);
        if (Math.min(...lengths) !== Math.max(...lengths)) {
            throw new Error(`Your attributes are not of the same length!`);
        }
    }
    bind(gl) {
        bindProgram(gl, this.program.program);
        for (const a of this.attributes) {
            bindBufferToAttribute(gl, a.location, a.value);
        }
        for (const u of this.uniforms) {
            bindValueToUniform(gl, u.location, u.type, u.value);
        }
        for (const t of this.textures) {
            bindTextureToUniform(gl, t.texture.texture, t.bindPoint, t.location);
        }
        if (this.index) {
            bindIndexBuffer(gl, this.index.index);
        }
    }
    render(gl, background, frameBuffer, viewport) {
        if (!frameBuffer) {
            bindOutputCanvasToFramebuffer(gl, viewport);
        }
        else {
            bindFramebuffer(gl, frameBuffer, viewport);
        }
        if (background) {
            clearBackground(gl, background);
        }
        if (this.index) {
            drawElements(gl, this.index.index);
        }
        else {
            const firstAttribute = this.attributes[0];
            drawArray(gl, firstAttribute.value);
        }
    }
    updateAttributeData(gl, variableName, newData) {
        const attribute = first(this.attributes, el => el.variableName === variableName);
        if (!attribute) {
            throw new Error(`No such attribute ${variableName} to be updated.`);
        }
        updateBufferData(gl, attribute.value, newData);
    }
    updateUniformData(gl, variableName, newData) {
        const uniform = first(this.uniforms, el => el.variableName === variableName);
        if (!uniform) {
            throw new Error(`No such uniform ${variableName} to be updated.`);
        }
        uniform.value = newData;
    }
    updateTextureData(gl, variableName, newImage) {
        const original = first(this.textures, t => t.variableName === variableName);
        if (!original) {
            throw new Error(`No such texture ${variableName} to be updated.`);
        }
        const newTextureObject = updateTexture(gl, original.texture, newImage);
        original.texture = newTextureObject;
    }
}
export class Framebuffer {
    constructor(gl, width, height) {
        const fb = createFramebuffer(gl);
        const fbTexture = createEmptyTexture(gl, width, height);
        const fbo = bindTextureToFramebuffer(gl, fbTexture, fb);
        this.fbo = fbo;
    }
}
export function renderLoop(fps, renderFunction) {
    const tDeltaTarget = 1000 * 1.0 / fps;
    let tDelta = tDeltaTarget;
    let tStart, tNow, tSleep;
    const render = () => {
        tStart = window.performance.now();
        renderFunction(tDelta);
        tNow = window.performance.now();
        tDelta = tNow - tStart;
        tSleep = Math.max(tDeltaTarget - tDelta, 0);
        setTimeout(() => {
            requestAnimationFrame(render);
        }, tSleep);
    };
    render();
}
export class Entity {
    constructor(program, attributes, uniforms, textures, updateFunction) {
        this.program = program;
        this.attributes = attributes;
        this.uniforms = uniforms;
        this.textures = textures;
        this.updateFunction = updateFunction;
    }
    update(tDelta) {
        this.updateFunction(tDelta, this.attributes, this.uniforms);
    }
}
export class Engine {
    constructor() {
        this.entities = [];
    }
    renderLoop(gl, fps) {
        setup3dScene(gl);
        const tDeltaTarget = 1000 * 1.0 / fps;
        let tStart, tNow, tDelta, tSleep;
        let currentShader = '';
        const render = () => {
            tStart = window.performance.now();
            // Part 1: allow objects to update their state
            for (const e of this.entities) {
                e.update(tDeltaTarget);
            }
            // Part 2: do the actual rendering work here
            clearBackground(gl, [.7, .7, .7, 1]);
            for (const e of this.entities) {
                if (e.program.id !== currentShader) {
                    bindProgram(gl, e.program.program);
                    currentShader = e.program.id;
                }
                for (const a of e.attributes) {
                    bindBufferToAttribute(gl, a.location, a.value);
                }
                for (const u of e.uniforms) {
                    bindValueToUniform(gl, u.location, u.type, u.value);
                }
                for (const t of e.textures) {
                    bindTextureToUniform(gl, t.texture.texture, t.bindPoint, t.location);
                }
                gl.drawArrays(gl.TRIANGLES, 0, e.attributes[0].value.vectorCount);
            }
            // Part 3: time-management
            tNow = window.performance.now();
            tDelta = tNow - tStart;
            tSleep = Math.max(tDeltaTarget - tDelta, 0);
            setTimeout(() => {
                requestAnimationFrame(render);
            }, tSleep);
        };
        render();
    }
    addEntity(entity) {
        this.entities.push(entity);
        this.sortEntities();
    }
    sortEntities() {
        this.entities.sort((a, b) => {
            return (a.program.id > b.program.id) ? 1 : -1;
        });
    }
}
