// @dlr-eoc/core-ui/esm2020/lib/core-ui.module.mjs
import { NgModule } from '@angular/core';
import { GlobalAlertComponent } from './global-alert/global-alert.component';
import { AlertService } from './global-alert/alert.service';
import { GlobalProgressComponent } from './global-progress/global-progress.component';
import { ProgressService } from './global-progress/progress.service';
import { CommonModule } from '@angular/common';
import { ClarityModule } from '@clr/angular';
import { HeaderComponent } from './header/header.component';
import { ViewRefDirective, DynamicComponentComponent } from './dynamic-component/dynamic-component.component';
import * as i0 from "@angular/core";
export class CoreUiModule {
}
CoreUiModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: CoreUiModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CoreUiModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: CoreUiModule, declarations: [GlobalAlertComponent, GlobalProgressComponent, HeaderComponent, ViewRefDirective, DynamicComponentComponent], imports: [CommonModule, ClarityModule], exports: [GlobalAlertComponent, GlobalProgressComponent, HeaderComponent, ViewRefDirective, DynamicComponentComponent] });
CoreUiModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: CoreUiModule, providers: [AlertService, ProgressService], imports: [[CommonModule, ClarityModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: CoreUiModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [GlobalAlertComponent, GlobalProgressComponent, HeaderComponent, ViewRefDirective, DynamicComponentComponent],
                    imports: [CommonModule, ClarityModule],
                    exports: [GlobalAlertComponent, GlobalProgressComponent, HeaderComponent, ViewRefDirective, DynamicComponentComponent],
                    providers: [AlertService, ProgressService]
                }]
        }] });
