// @dlr-eoc/services-map-state/esm2020/lib/types/map-state.mjs
export class MapState {
    constructor(zoom, center, options, extent = [-180.0, -90.0, 180.0, 90.0], time = new Date().toISOString()) {
        const defaultOptions = {
            maxzoom: 0,
            minzoom: 0,
            notifier: 'map'
        };
        this.zoom = zoom;
        this.center = center;
        this.extent = extent;
        this.time = time;
        this.options = Object.assign(defaultOptions, options);
    }
    sameCenter(center) {
        if (this.center.lat === center.lat && this.center.lon === center.lon) {
            return true;
        }
        else {
            return false;
        }
    }
    sameZoom(zoom) {
        if (this.zoom === zoom) {
            return true;
        }
        else {
            return false;
        }
    }
    sameExtent(extent) {
        const len = extent.length;
        let isSame = false;
        if (this.extent.length === len) {
            isSame = this.extent.every((v, i) => extent[i] === v);
        }
        return isSame;
    }
    sameNotifier(notifier) {
        if (this.options.notifier === notifier) {
            return true;
        }
        else {
            return false;
        }
    }
}
